Tracing GC for konserve

csm opened the pull request:
Based partially on https://github.com/datacrypt-project/hitchhiker-tree/pull/24. Rewritten to primarily use core.async.

Would need something like https://github.com/replikativ/konserve/pull/29 to iterate keys in the store.

* src/hitchhiker/tree/konserve.cljc (create-id): new function; prepends
  the current timestamp as hex to the UUID key.
  (KonserveBackend.-write-node): use create-id to generate the storage ID.
* src/hitchhiker/tree/tracing-gc/konserve.cljc: new namespace.
* src/hitchhiker/tree/tracing-gc.cljc: new namespace.
* .gitignore: ignore IntelliJ files.
* project.clj: update konserve to 0.6.0-SNAPSHOT.

noonhomie committed:
Tracing GC.

Based partially on https://github.com/datacrypt-project/hitchhiker-tree/pull/24
Rewritten to primarily use core.async.

* src/hitchhiker/tree/konserve.cljc (create-id): new function; prepends
  the current timestamp as hex to the UUID key.
  (KonserveBackend.-write-node): use create-id to generate the storage ID.
* src/hitchhiker/tree/tracing-gc/konserve.cljc: new namespace.
* src/hitchhiker/tree/tracing-gc.cljc: new namespace.
* .gitignore: ignore IntelliJ files.
* project.clj: update konserve to 0.6.0-SNAPSHOT.

noonhomie committed:
Remove logging (don't know if it is available for cljs or not).

noonhomie committed:
Finish the thought in create-id docstring.

whilo commented on src/hitchhiker/tree/bootstrap/konserve.cljc:
```diff
@@ -53,6 +53,16 @@
+  the tree anymore, but also skips keys that were added after the GC
+  process begins."
+  [uuid]
+  (format "%016x.%s" (System/currentTimeMillis) uuid))
```
I would prefer not to use custom string serialization, but plain Clojure datastructures, i.e. `[(System/currentTimeMillis) uuid]`. For many backends konserve internally uses hasch to coerce the key again to a uuid anyway and for the ones where we still store the serialized key it would not be longer in serialized form than this string serialization.

whilo replied:
Maybe actually using an instance data type (java.util.Date) would be better than an implicit long for milliseconds.

csm replied:
That sounds totally fine. The original intent with using strings was to have IDs that could sort properly, but I did away with that in the konserve patch anyway, so just a vector of `[instant guid]` works fine too -- I'll make that change.

csm replied:
Also given the discussion in datahike over forking databases, the ID could be some kind of vector clock (say, an interval tree clock) instead of a simple timestamp. The stamping mechanism can even be abstracted into a protocol, so different mechanisms could be plugged in.

whilo replied:
Yes, that is why I had mentioned the lattices. Interval Tree Clocks are a good idea. I know the original paper, https://haslab.uminho.pt/cbm/files/itc.pdf. What references have you taken a look at?

whilo commented on src/hitchhiker/tree/tracing_gc.cljc:
```diff
@@ -0,0 +1,50 @@
+   all-keys should be a core.async channel that will contain every key in storage.
+   delete-fn will be called on every key that should be deleted during the sweep phase. It is expected to return a channel that yields when the item is deleted."
+  [gc-scratch gc-roots all-keys delete-fn]
+  (let [mark-phase (async/go-loop [roots gc-roots]
```
I would mark node ids purely in memory. Besides the slow down of writing the marked attribute to the files, storing the marks will cause problems for crashes during GC and will leave stale files behind. I also think at the moment all files that have been marked once will never be garbage collected again. This also should yield a simpler algorithm that is easier to reason about. A simple set of marked nodes should be sufficient, right?

csm replied:
The design here just backs marked nodes with a konserve instance, and one can easily just back it with a memory store.

I thought having the node marking backend be configurable would be a good design; that way the user can choose whether they can do in-memory, or if they need a file-backed store.

whilo replied:
That makes sense, I was just wondering when you would want to store marks on disk. I can think of having a parallel GC sweep with core.async (konserve's filestore at least has non-blocking reads on the JVM) at some point. But even then you would not need to durably store the marks. Storing each mark in a separate key would definitely be very slow for IO backends.

whilo replied:
I am fine with merging the current version if you think it makes sense to persist the marks. I have just so far always tried to keep the abstractions minimal so that it is easier to adapt the replikativ libraries when somebody has the need.

whilo reviewed (changes requested):
Your work looks very promising and it is great fun to implement my first GC with you! I think we can get that done soon and plug it into Datahike. Could you also write some tests for the GC code?

noonhomie committed:
Use [stamp guid] as konserve key.

Also add protocol for custom stamping mechanisms, just use a current
timestamp for stamping nodes by default.

* src/hitchhiker/tree/bootstrap/konserve.cljc
  (PKonserveStampGenerator): new protocol.
  (KonsorveTimestampGenerator): new record.
  (+default-stamp-generator+): new var.
  (create-id): take a stamp generator, use that to create the timestamp.
  Return a vector of `[stamp guid]`, not a string.
  (KonserveAddr): implement -raw-address method.
* src/hitchhiker/tree/bootstrap/redis.clj (RedisAddr): implement -raw-address
  method.
* src/hitchhiker/tree/tracing_gc/konserve.cljc (compare-stamps): new fn,
  implement for clj and cljs (do the right thing in cljs).
  (within-epoch?): rename to after-epoch?. Treat addresses as vectors
  of timestamp, guid.
* src/hitchhiker/tree/utils/plot.clj: pass nil for stamp-generator arg
  for ->KonsorveBackend.
* src/hitchhiker/tree/node.cljc (IAddress): add -raw-address method.
* src/hitchhiker/tree/tracing-gc.cljc (trace-gc!): use -raw-address to
  get the underlying storage address.
* src/hitchhiker/tree.cljc (IndexNode, DataNode): implement -raw-address
  method.
* test/hitchhiker/tree/node/testing.cljc (TestingAddr): implement -raw-address
  method.
* test/hitchhiker/tree/testing_gc_test.clj: new test.
* test/hitchhiker/konserve_test.cljc: pass nil as stamp-generator to
  ->KonserveBackend.

noonhomie committed:
Just mark GC nodes in-memory.

Rename konserve -> epoch, it better captures what that does now.
Don't make tracing-gc! async.
Update gc test.

whilo commented:
That is already considerably simpler. It would be ideal to collect all hitchhiker-tree marks in a set and then return it. That way we can add additional keys to mark (e.g. for the roots of all Datahike branches or maybe other things in store) and call a `sweep` function separately with the marks set. Sorry, I should have been more explicit about this. It would also be nice if the collection of marks would work with core.async as well (without `<??`), but we can fix that later.

whilo commented:
The mark phase is a pure function because the tree roots we are giving to it are constants and the store is immutable then. We just happen to read constants from disk. That way we separate it from the side-effects of removing the stale nodes.

noonhomie committed:
Separate mark and sweep!

mark just collects addresses into a set (possibly collecting more
addresses into an existing set).

sweep! walks a sequence (not channel) of addresses, and calls delete-fn
on any address that is both accepted by accept-fn and does not appear
in the marked addresses set.

Removes IGCScratch protocol.

epoch just contains a function-builder-function that returns true if
the timestamp in a key is before an epoch.

noonhomie committed:
Simplify further.

Don't take any accept-fn in sweep!, just leave it up to the caller
to filter the seq of addresses.

Leave hitchhiker-tree.tracing-gc.epoch for accept-before-epoch fn,
since it's useful for consumers of the GC API.

whilo commented on src/hitchhiker/tree/tracing_gc.cljc:
```diff
@@ -4,45 +4,40 @@
+  Returns all marked addresses."
+  ([gc-roots] (mark gc-roots #{}))
+  ([gc-roots addresses]
+   (loop [addresses (volatile! (transient addresses))
```
Ok, one last thing, we do not need to use a mutable object for addresses and the loop here. I would just use a recursion and pure values. Actually we can even do it along the tree without the loop and merge the resulting sets of marks on the way up again, there is no risk of a stackoverflow. 

(Later we can collect go channels that return either marks or nothing and we can pipeline the reads in parallel that way.)

whilo replied:
Forget the pipelining comment, it is a bit trickier than that.

csm replied:
I'm unsure what the remaining issue is with this implementation. If you would prefer it written some other way, feel free to change it yourself.

whilo replied:
Ok, sure.

noonhomie committed:
Update ns docs.

whilo commented:
@csm Hey. We have now implemented this with the help of metadata support in our key-value store protocol, which does not clutter the value semantics with timestamps. Thanks a lot for pushing the GC and if you have any further needs, feel free to reach out or open more PRs.
